clusterloader2/pkg/measurement/common/wait_for_controlled_pods.go

```diff
@@ -24,123 +24,380 @@ package common
+
+	switch action {
+	case "start":
+		w.kind, err = util.GetString(config.Params, "kind")
+		if err != nil {
+			return summaries, err
+		}
+		w.namespace, err = util.GetStringOrDefault(config.Params, "namespace", metav1.NamespaceAll)
+		if err != nil {
+			return summaries, err
+		}
+		w.labelSelector, err = util.GetStringOrDefault(config.Params, "labelSelector", "")
+		if err != nil {
+			return summaries, err
+		}
+		w.fieldSelector, err = util.GetStringOrDefault(config.Params, "fieldSelector", "")
+		if err != nil {
+			return summaries, err
+		}
+		w.operationTimeout, err = util.GetDurationOrDefault(config.Params, "operationTimeout", defaultOperationTimeout)
+		if err != nil {
+			return summaries, err
+		}
+		return summaries, w.start(config.ClientSet)
+	case "gather":
+		syncTimeout, err := util.GetDurationOrDefault(config.Params, "syncTimeout", defaultWaitForRCPodsTimeout)
+		if err != nil {
+			return summaries, err
+		}
+		return summaries, w.gather(config.ClientSet, syncTimeout)
+	case "stop":
+		w.stop()
+		return summaries, nil
+	default:
+		return summaries, fmt.Errorf("unknown action %v", action)
+	}
+}
+
+func (w *waitForControlledPodsRunningMeasurement) start(c clientset.Interface) error {
+	if w.informer != nil {
+		glog.Infof("%v: wait for controlled pods measurement already running", waitForControlledPodsRunningName)
+		return nil
+	}
+	glog.Infof("%v: starting wait for controlled pods measurement...", waitForControlledPodsRunningName)
+	optionsModifier := func(options *metav1.ListOptions) {
+		options.FieldSelector = w.fieldSelector
+		options.LabelSelector = w.labelSelector
+	}
+	listerWatcher := cache.NewFilteredListWatchFromClient(c.CoreV1().RESTClient(), w.kind+"s", w.namespace, optionsModifier)
+	w.isRunning = true
+	w.stopCh = make(chan struct{})
+	w.informer = cache.NewSharedInformer(listerWatcher, nil, 0)
+	w.informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			addF := func() {
+				w.handleObject(c, nil, obj)
+			}
+			w.queue.Add(&addF)
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			updateF := func() {
+				w.handleObject(c, oldObj, newObj)
+			}
+			w.queue.Add(&updateF)
+		},
+		DeleteFunc: func(obj interface{}) {
+			deleteF := func() {
+				w.handleObject(c, obj, nil)
+			}
+			w.queue.Add(&deleteF)
+		},
+	})
+
+	for i := 0; i < waitForControlledPodsWorkers; i++ {
+		w.workerGroup.Start(w.worker)
+	}
+	go w.informer.Run(w.stopCh)
+	timeoutCh := make(chan struct{})
+	timeoutTimer := time.AfterFunc(informerSyncTimeout, func() {
+		close(timeoutCh)
+	})
+	defer timeoutTimer.Stop()
+	if !cache.WaitForCacheSync(timeoutCh, w.informer.HasSynced) {
+		return fmt.Errorf("timed out waiting for caches to sync")
+	}
+
+	return nil
+}
+
+func (w *waitForControlledPodsRunningMeasurement) worker() {
+	for {
+		f, stop := w.queue.Get()
+		if stop {
+			return
+		}
+		funcF := f.(*func())
+		(*funcF)()
```

Comment:
(*f.(func())()

[if you're not handling casting errors anyway]

Reply:
Done.